tools/kagami/src/crypto.rs

```diff
@@ -1,14 +1,135 @@
+    /// Path to private key to to sign genesis block
+    #[clap(long, group = "private_key")]
+    private_key_path: Option<PathBuf>,
+    /// Public key of the corresponding private key
+    #[clap(long, group = "public_key")]
+    public_key_string: Option<String>,
+    /// Path to public key of the corresponding private key
+    #[clap(long, group = "public_key")]
+    public_key_path: Option<PathBuf>,
+    /// Path to json-serialized keypair
+    #[clap(long, short, group = "private_key", group = "public_key")]
+    keypair_path: Option<PathBuf>,
+    /// Unique id of blockchain
+    #[clap(long)]
+    chain_id: ChainId,
+    /// Path to genesis json file
+    #[clap(long, short)]
+    genesis_path: PathBuf,
+    /// Path to signed genesis output file
+    #[clap(long, short, group = "format")]
+    output_path: Option<PathBuf>,
+}
+
+impl GenesisSigningArgs {
+    fn get_private_key(&self) -> Result<PrivateKey> {
+        let private_key_bytes =
+            Self::get_key_bytes(&self.private_key_path, &self.private_key_string)?;
+        PrivateKey::from_bytes(self.algorithm.0, private_key_bytes.borrow()).wrap_err_with(|| {
+            eyre!(
+                "Failed to parse private key for algorithm `{}`",
+                self.algorithm
+            )
+        })
+    }
+
+    fn get_public_key(&self) -> Result<PublicKey> {
+        let public_key_bytes = Self::get_key_bytes(&self.public_key_path, &self.public_key_string)?;
+        PublicKey::from_bytes(self.algorithm.0, public_key_bytes.borrow()).wrap_err_with(|| {
+            eyre!(
+                "Failed to parse public key for algorithm `{}`",
+                self.algorithm
+            )
+        })
+    }
+
+    fn get_key_bytes<'a, P: AsRef<Path>>(
+        path: &Option<P>,
+        value: &'a Option<String>,
+    ) -> Result<Cow<'a, [u8]>, std::io::Error> {
+        match (path, value) {
+            (Some(path_buf), None) => Ok(Cow::Owned(fs::read(path_buf)?)),
+            (None, Some(hex)) => Ok(Cow::Borrowed(hex.as_bytes())),
+            _ => unreachable!("Clap group invariant"),
+        }
+    }
+
+    fn read_keypair<P: AsRef<Path>>(path: P) -> Result<KeyPair> {
+        let bytes = fs::read(path)?;
+        Ok(serde_json::from_slice(bytes.as_slice())?)
+    }
+}
+
+impl<T: Write> RunArgs<T> for GenesisSigningArgs {
+    fn run(self, writer: &mut BufWriter<T>) -> Outcome {
+        let key_pair = if let Some(path) = self.keypair_path {
+            Self::read_keypair(path)?
+        } else {
+            let (public_key, private_key) = (self.get_public_key()?, self.get_private_key()?);
+            KeyPair::new(public_key, private_key)?
+        };
+
+        let genesis_block = RawGenesisBlock::from_path(&self.genesis_path)?;
+        let encoded_genesis_network =
+            GenesisNetwork::new(genesis_block, &self.chain_id, &key_pair).encode();
+
+        if let Some(path) = self.output_path {
+            fs::write(&path, encoded_genesis_network)?;
```

Comment:
So, we take a JSON genesis file, and write a binary SCALE-encoded file back.

What are other options, should we consider them? E.g. making JSON output so that it is readable.

Reply:
I would prefer to be as simple as possible here.

Reply:
Added an option to output signed genesis block as json to stdout

Reply:
either we output signed genesis as JSON or we output it as SCALE. IMHO JSON is better

Reply:
I don't like the idea that the final configuration of the peer will consist of the SCALE encoded genesis blob. Instead, I think, it should include the original untouched `genesis.json` with an external file containing signatures.

For example, `kagami crypto sign-genesis` (or `kagami genesis sign`) command might get a `genesis.json` as input and produce `genesis-signatures.json` as the output. It will contain an array of hexes - signatures of each genesis transaction:

```json
["<hex for tx1 (executor)>", "<hex for tx2>", "<...>"]
```

Then, this file with signatures must be specified in the config for the peer started with `--submit-genesis` arg:

```toml
[genesis]
public_key = "..."
block_file = "genesis.json"
signatures_file = "genesis-signatures.json"
```

Reply:
> I don't like the idea that the final configuration of the peer will consist of the SCALE encoded genesis blob. Instead, I think, it should include the original untouched genesis.json with an external file containing signatures.

we can agree that it shouldn't be a SCALE encoded blob. But why take 2 files in the configuration when you can take only 1 containing JSON serialized signed genesis transaction instead of 

> signatures of each genesis transaction:

there should be exactly 1 genesis transaction to avoid the possibility of malicious peer removing or reordering some transactions. Or we can have a fixed amount of genesis transactions where position of each is hardcoded. For example, we can expect 1st transaction to be executor and 2nd to be everything else

Reply:
> But why take 2 files in the configuration when you can take only 1 containing JSON serialized signed genesis transaction instead of

it's not a bad idea, I just wanted genesis to have the layout of `SignedTransaction` because of sdk issues

Reply:
> there should be exactly 1 genesis transaction

Indeed, I forgot it. So it's even simpler: we need produce a single hex.

> But why take 2 files in the configuration when you can take only 1 containing JSON serialized signed genesis transaction instead

It is confusing in many minor practical aspects:

- you produce almost a copy of the original file, but with a small addition
- if you deploy Iroha (i.e. you deal with all this signing), you start to bother about having both copies, storing the original one somewhere
- oh, I messed something, let me re-sign the block with a different key! Wait.. can I re-sign the signed block, or only the empty one? Let me find the original one, where is it...

Actually, to make things **really** simple:

```toml
[genesis]
public_key = "<hex>"
file = "genesis.json"
signature = "<hex>"
```

Reply:
Added `encoded_config` parameter that contains hex string which encodes 3 parameters: `chain_id`, `creation_time` and `signatures`